### PR TITLE
952 layout title

### PR DESCRIPTION
**Pull request by jenlampton**
Fixes https://github.com/backdrop/backdrop-issues/issues/952
Replaces https://github.com/backdrop/backdrop/pull/1234/commits

### CODE REVIEW

**Comment by jenlampton** [comment]
Comment from previous PR: 

> The "Displaying: Default title" status on the "Page title" pseudo-region is really nice, but it doesn't seem to be updated upon form submitting the dialog (the "dirty form" flag is fired though) - only after saving the layout and then re-editing is one able to view the change.

It looks like there is already code in here to solve thls..  

``` php
'#ajax' => array(
      'callback' => 'layout_ajax_form_save_dialog',
    ),
```

But for some reason it's not doing what i think it should be doing.  Still digging deeper...

**Comment by jenlampton** [comment]
I think `layout_ajax_form_save_dialog` might be too specific for our use here, I think we may need a custom ajax command to update the title information for us.  going to try that...

**Comment by jenlampton** [comment]
I got it working! :) Pushing a new PR now.

**Comment by klonos** [comment]
@jenlampton can you please force this so we can get a PR sandbox? I'd :heart: this to land in 1.4.0

**Comment by jenlampton** [comment]
forced an update @klonos sorry for the delay.

**Comment by klonos** [comment]
No worries @jenlampton I don't have much time to test extensively anyways. Perhaps I'll manage to allocate some more time throughout this weekend.

Initial comments/thoughts:
- It looks so much better having the page title as a pseudo-region :+1: to the whole thing!
- Not sure if we should be rendering this above the main content region or above the rest of the regions (where we initially had the respective drop-down menu). Leaning towards above the main content.
- There is enough room to be more verbose about the currently selected configuration (like including the help text for each option for example). Will need to give this more thought though.
- The help text for the options could be improved. More on that when I have some time.

...in general though, this rocks!! We should get it in for 1.4.0.

**Comment by jenlampton** [comment]
> Not sure if we should be rendering this above the main content region or above the rest of the regions (where we initially had the respective drop-down menu).

It will depend on the layout, as each one has a "title" location defined, and we should put it right there :)

File: core/modules/layout/layout.module [hunk 1]
```diff
@@ -1549,3 +1555,36 @@ function layout_set_handler_contexts(array &$contexts, array $all_contexts, arra
...
+/**
+ * Gets the title description for the current layout.
+ *
+ * @param $layout
+ *   Layout object.
+ *
+ * @return $description
+ *   Description of title source.
+ */
+function layout_get_title_description($layout) {
```
Comment (quicksketch) [inline core/modules/layout/layout.module]: Let's start adding code hints for all new code. `$layout` should be `Layout $layout` indicating that only `Layout` objects should be allowed to be passed in here. It also helps with auto-complete within IDEs.

File: core/modules/layout/layout.module [hunk 2]
```diff
@@ -1549,3 +1555,36 @@ function layout_set_handler_contexts(array &$contexts, array $all_contexts, arra
   // All required contexts populated successfully.
   return TRUE;
 }
+
+/**
+ * Gets the title description for the current layout.
+ *
+ * @param $layout
+ *   Layout object.
```
Comment (quicksketch) [inline core/modules/layout/layout.module]: As below, let's code hint the PHPdoc here. `@param $layout` can be `@param Layout $layout`. Even though it seems redundant to human readers, it helps with IDE completion.

File: core/modules/layout/layout.module (hunk 1 above)
Comment (jenlampton) [inline core/modules/layout/layout.module]: done.

File: core/modules/layout/layout.module (hunk 2 above)
Comment (jenlampton) [inline core/modules/layout/layout.module]: done.

**Comment by jenlampton** [comment]
Rebased, and added new tests.

**Comment by klonos** [comment]
```

    Notice: Undefined index: header in bartik_preprocess_layout() (line 35 of /home/qa/github/backdrop/backdrop/pr/1284/core/themes/bartik/template.php).
    Notice: Undefined index: header in include() (line 28 of /home/qa/github/backdrop/backdrop/pr/1284/core/modules/layout/templates/layout.tpl.php).
    Notice: Undefined index: top in include() (line 34 of /home/qa/github/backdrop/backdrop/pr/1284/core/modules/layout/templates/layout.tpl.php).
    Notice: Undefined index: content in include() (line 64 of /home/qa/github/backdrop/backdrop/pr/1284/core/modules/layout/templates/layout.tpl.php).
    Notice: Undefined index: footer in include() (line 68 of /home/qa/github/backdrop/backdrop/pr/1284/core/modules/layout/templates/layout.tpl.php).
    Notice: Undefined index: default region in LayoutRendererStandard->ensurePageContentBlock() (line 549 of /home/qa/github/backdrop/backdrop/pr/1284/core/modules/layout/plugins/renderers/layout_renderer_standard.inc).
    Notice: Undefined index: regions in LayoutRendererStandard->prepareRegions() (line 284 of /home/qa/github/backdrop/backdrop/pr/1284/core/modules/layout/plugins/renderers/layout_renderer_standard.inc).
    Warning: Invalid argument supplied for foreach() in LayoutRendererStandard->prepareRegions() (line 284 of /home/qa/github/backdrop/backdrop/pr/1284/core/modules/layout/plugins/renderers/layout_renderer_standard.inc).

```

**Comment by quicksketch** [comment]
This needs to be updated again after the layout terminology changes.

**Comment by quicksketch** [comment]
Pushed a revised PR that resolves current merge conflicts. This looks really good!

**Comment by quicksketch** [comment]
Close/reopen for new sandbox.

**Comment by quicksketch** [comment]
Updated the PR to assign @docwilmot credit for this work as the original author. This looks fantastic. Merged into 1.x for 1.5.0. Thanks @docwilmot and @jenlampton!

**Pull request merged**